OIDC: Make scopes configurable

markylaing opened the pull request:
In the end, just having `oidc.scopes` was the way to go (see https://github.com/canonical/lxd/pull/14888#issuecomment-2642783581). I've added tests for the validation of the config key and also tested it locally with Keycloak by disabling the `profile` claim.

Closes #14141.

markylaing commented on client/lxd_oidc.go:
```diff
@@ -115,11 +115,22 @@ func (o *oidcClient) do(req *http.Request) (*http.Response, error) {
 	issuer := resp.Header.Get("X-LXD-OIDC-issuer")
 	clientID := resp.Header.Get("X-LXD-OIDC-clientid")
 	audience := resp.Header.Get("X-LXD-OIDC-audience")
```
@tomponline I've never liked that we use headers for this. The API documentation for the [`/1.0?public` endpoint](https://documentation.ubuntu.com/lxd/en/latest/api/#/server/server_get_untrusted) says:

> Shows a small subset of the server environment and configuration which is required by untrusted clients to reach a server.

IMO we should have a small subset of config available over /1.0 for untrusted clients, including these values.

simondeziel commented on doc/api-extensions.md:
```diff
@@ -2505,6 +2505,7 @@ And lastly, adds a `project` field on leases, leases can be retrieved via `/1.0/
 
 ## `network_ovn_uplink_vlan`
 
+<<<<<<< HEAD
```
Conflict resolution leftover ;)

markylaing replied:
Whoopsie!

simondeziel commented on lxd/auth/oidc/oidc.go:
```diff
@@ -453,12 +463,7 @@ func (o *Verifier) setRelyingParty(ctx context.Context, host string) error {
-	}
-
-	relyingParty, err := rp.NewRelyingPartyOIDC(ctx, o.issuer, o.clientID, "", fmt.Sprintf("https://%s/oidc/callback", host), oidcScopes, options...)
+	relyingParty, err := rp.NewRelyingPartyOIDC(ctx, o.issuer, o.clientID, "", fmt.Sprintf("https://%s/oidc/callback", host), o.scopes, options...)
```
```suggestion
	relyingParty, err := rp.NewRelyingPartyOIDC(ctx, o.issuer, o.clientID, "", "https://"+host+"/oidc/callback", o.scopes, options...)
```

markylaing replied:
Fixed.

simondeziel commented on test/suites/oidc.sh:
```diff
@@ -4,6 +4,14 @@ test_oidc() {
+  # Check OIDC scopes validation
+  ! lxc config set oidc.scopes "my-scope" || false # Doesn't contain "email" or "openid"
+  ! lxc config set oidc.scopes "my-scope email" || false # Doesn't contain "openid"
+  ! lxc config set oidc.scopes "my-scope openid" || false # Doesn't contain "openid"
```
```suggestion
  ! lxc config set oidc.scopes "my-scope openid" || false # Doesn't contain "email"
```

markylaing replied:
Fixed.

minaelee commented on doc/api-extensions.md:
```diff
@@ -2599,3 +2600,10 @@ The following pool level configuration keys have been added:
+
+## `oidc_scopes`
+
+This API extension enables setting an {config:option}`server-oidc:oidc.scopes` configuration key which accepts a space separated list of OIDC scopes to request from the identity provider.
```
```suggestion
This API extension enables setting an {config:option}`server-oidc:oidc.scopes` configuration key, which accepts a space-separated list of OIDC scopes to request from the identity provider.
```

markylaing replied:
Fixed, thanks :)

minaelee commented on lxd/cluster/config/config.go:
```diff
@@ -679,11 +680,33 @@ var ConfigSchema = config.Schema{
+	// ---
+	//  type: space-delimited string
+	//  scope: global
+	//  shortdesc: Space separated list of OpenID Connect scopes
```
```suggestion
	//  shortdesc: Space-separated list of OpenID Connect scopes
```

markylaing replied:
Fixed.

markylaing commented:
@tomponline this is ready when you have time. Thanks.

tomponline commented on client/lxd_oidc.go:
@markylaing as discussed in order to maintain backwards and forwards compat between client and server, lets:

1. Have the server continue to send the `X-LXD-OIDC-groups-claim` header - so that old clients continue to request this as a scope.
2. Have the new client detect that the server supports the new API extension and disregard the `X-LXD-OIDC-groups-claim` header for scopes and instead use `X-LXD-OIDC-scopes`.

Update commit to remove references to `X-LXD-OIDC-additional-scopes` too please.

markylaing replied:
I've implemented this now.

markylaing commented:
@tomponline ready for another round when you're ready. I realised that I forgot to cherry-pick the patch from #14888 so I've added it now and also added a unit test for that.

tomponline commented on lxd/patches.go:
```diff
@@ -1445,4 +1447,40 @@ DELETE FROM storage_pools_config WHERE key = "powerflex.sdt"
+// The oidc.groups.claim value was initially added to scopes but shouldn't have been. This patch will allow users with
+// working identity provider group mappings to continue using them by continuing to request the claim as an additional
+// scope.
+func patchOIDCGroupsClaimScope(_ string, d *Daemon) error {
```
@markylaing this function will be run on each cluster member, at different times.

I notice that it is loading the cluster config from the DB inside the function, and then modifying the global config.

But is it safe when potentially this function is run again on a different cluster member after the global config has been modified by this function at an earlier stage?

markylaing replied:
Ahh of course :facepalm: will fix

markylaing replied:
Fixed. Thanks for catching :)

tomponline reviewed (commented):
one comment
